### PR TITLE
iox-#1884 Use inequality operator of underlying type in optional ineq…

### LINKED ISSUE #1884: Optional inequality implemented with equality operator of type.
## Required information

**Observed result or behaviour:**
A type which does not have an equality operator implemented but only an inequality operator cannot be tested for inequality in an optional since it is implemented with the equality operator.

See this code example
```cxx
struct A {
  bool operator!=(const A &) { return true; }
};

iox::optional<A> a, b;
a != b; // compiler failure
```

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1884

### CODE REVIEW

**Review by elBoberido — APPROVED** [review]
Suggested a more efficient version but if you do not agree we can merge anyway

File: iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl [hunk 1]
```diff
@@ -306,7 +306,10 @@ bool operator==(const optional<T>& lhs, const optional<T>& rhs) noexcept
 template <typename T>
 bool operator!=(const optional<T>& lhs, const optional<T>& rhs) noexcept
 {
-    return !(lhs == rhs);
+    const auto onlyLhsNul = !lhs.has_value() && rhs.has_value();
+    const auto onlyRhsNul = lhs.has_value() && !rhs.has_value();
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl:310]: Why not `onlyLhsNull` and `onlyRhsNull`?

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

File: iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl (hunk 1 above)
Comment (elfenpiff) [reply iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl:310]: When one of the values is null and the other is not they are not equal but if they both contain a value we have to delegate this to the inequality operator of the type.

Does it help?!

File: iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl [hunk 2]
```diff
@@ -306,7 +306,10 @@ bool operator==(const optional<T>& lhs, const optional<T>& rhs) noexcept
...
+    const auto bothValuesUnequal = (lhs.has_value() && rhs.has_value()) && (*lhs != *rhs);
+    return bothValuesUnequal || onlyRhsNul || onlyLhsNul;
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl:312]: @FerdinandSpitzschnueffler @elBoberido I didn't believe you so I played around with godbolt to see if there is any difference. And yes the gcc compiler seem to optimize your solution a bit better (but this is just a naive observation from my side by glimpsing at the assembler code). Here the link: https://godbolt.org/z/d9n8qc96W

[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl:312]: Oh another fun fact ... the current clang compiler optimizes the code even better and there it does not make any difference!

So it seems that your suggestion is an improvement for current and older gcc compilers. But clang seem able to handle it.

With this in mind I would optimize for simplicity and readability and would keep the code as it is.

File: iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl (hunk 1 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl:310]: @elfenpiff I was referring to the names. I wondered why there is no second `l` in `*Nul`. Sorry, I should have phrased that more clearly.

Comment (elfenpiff) [reply iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl:310]: @FerdinandSpitzschnueffler ...mmmhhh I think I am lazy ;)

**Pull request merged**